File: libs/go-platform/envelope/envelope.go
// Package envelope defines the versioned event envelope every Kafka message on
// the platform uses. Payloads are service-specific and validated by the
// consumer; the envelope itself is stable and shared.
package envelope

import (
	"encoding/json"
	"errors"
	"time"

	"github.com/google/uuid"
)

// Envelope wraps a domain event with routing and tracing metadata.
type Envelope struct {
	EventID       string          `json:"event_id"`
	EventType     string          `json:"event_type"`
	EventVersion  int             `json:"event_version"`
	OccurredAt    time.Time       `json:"occurred_at"`
	Producer      string          `json:"producer"`
	CorrelationID string          `json:"correlation_id"`
	CausationID   string          `json:"causation_id"`
	Payload       json.RawMessage `json:"payload"`
}

// New builds an envelope for a payload that will be JSON-encoded.
func New(eventType string, version int, producer, correlationID, causationID string, payload any) (Envelope, error) {
	raw, err := json.Marshal(payload)
	if err != nil {
		return Envelope{}, err
	}
	if correlationID == "" {
		correlationID = uuid.NewString()
	}
	return Envelope{
		EventID:       uuid.NewString(),
		EventType:     eventType,
		EventVersion:  version,
		OccurredAt:    time.Now().UTC(),
		Producer:      producer,
		CorrelationID: correlationID,
		CausationID:   causationID,
		Payload:       raw,
	}, nil
}

// Validate checks the required fields are populated and well-formed.
func (e Envelope) Validate() error {
	switch {
	case e.EventID == "":
		return errors.New("envelope: event_id is required")
	case e.EventType == "":
		return errors.New("envelope: event_type is required")
	case e.EventVersion < 1:
		return errors.New("envelope: event_version must be >= 1")
	case e.OccurredAt.IsZero():
		return errors.New("envelope: occurred_at is required")
	case e.Producer == "":
		return errors.New("envelope: producer is required")
	case len(e.Payload) == 0:
		return errors.New("envelope: payload is required")
	}
	if _, err := uuid.Parse(e.EventID); err != nil {
		return errors.New("envelope: event_id must be a UUID")
	}
	return nil
}

// Decode unmarshals the payload into v.
func (e Envelope) Decode(v any) error { return json.Unmarshal(e.Payload, v) }

// Bytes returns the JSON encoding of the envelope.
func (e Envelope) Bytes() []byte {
	b, _ := json.Marshal(e)
	return b
}

// Parse decodes and validates an envelope from bytes.
func Parse(b []byte) (Envelope, error) {
	var e Envelope
	if err := json.Unmarshal(b, &e); err != nil {
		return Envelope{}, err
	}
	return e, e.Validate()
}

File: libs/py-common/auralis_common/envelope.py
"""Versioned Kafka event envelope, matching the Go platform definition."""

from __future__ import annotations

import uuid
from datetime import UTC, datetime
from typing import Any

from pydantic import BaseModel, Field


def _utcnow() -> datetime:
    return datetime.now(UTC)


class Envelope(BaseModel):
    """Wraps a domain event with routing and tracing metadata."""

    event_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    # Required-field checks mirror the Go envelope.Validate(): a message that the
    # Go platform would reject must not be silently accepted here either.
    event_type: str = Field(min_length=1)
    event_version: int = Field(default=1, ge=1)
    occurred_at: datetime = Field(default_factory=_utcnow)
    producer: str = Field(min_length=1)
    correlation_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    causation_id: str = ""
    payload: dict[str, Any] = Field(min_length=1)

    def to_bytes(self) -> bytes:
        return self.model_dump_json().encode("utf-8")

    @classmethod
    def from_bytes(cls, raw: bytes) -> Envelope:
        return cls.model_validate_json(raw)


def new_envelope(
    event_type: str,
    payload: dict[str, Any],
    *,
    producer: str,
    version: int = 1,
    correlation_id: str | None = None,
    causation_id: str = "",
) -> Envelope:
    return Envelope(
        event_type=event_type,
        event_version=version,
        producer=producer,
        correlation_id=correlation_id or str(uuid.uuid4()),
        causation_id=causation_id,
        payload=payload,
    )
